### DISCUSSION TITLE
How to check if a tree sequence has "dangling nodes"

### DISCUSSION THREAD

**Discussion by hyanwong** [tskit-dev/tskit#2499] [Q&A]
I would find it useful to check whether a tree sequence has any leaf nodes that are not samples (i.e. "dangling" nodes). I guess I   could do this by iterating over all the leaves of all the trees, but that's probably very slow for large tree sequences. There must be a more efficient way to do this by looking at the edge table (especially if we ignore the problem of "dead branches": topology linking nonsample nodes that is not connected to any root)

```
def has_dangling(ts):
    for tree in ts.trees():
        for leaf in tree.leaves():
            if not ts.is_sample(leaf):
                return True
    return False
```

**Comment by petrelharp** [comment]
Wouldn't something like this do it? (i.e., just check if there's any non-sample nodes that aren't parents to any edge)
```
np.any(np.logical_and(
    np.bincount(tables.edges.parent, minlength=tables.nodes.num_rows),
    tables.nodes.flags & tskit.NODE_IS_SAMPLE
))
```

**Comment by jeromekelleher** [comment]
I think if the definition is exactly your function above @hyanwong there's not much else that can be done, since this is finding the leaves reachable from roots that are not samples. It's hard to see how to do this exactly (in all the possible awkward corner cases) by looking at the edge table.

A weaker definition could be done efficiently via something like @petrelharp has given above.
